cli/src/complete.rs

```diff
@@ -329,35 +333,116 @@ fn revisions(revisions: Option<&str>) -> Vec<CompletionCandidate> {
             .map_err(user_error)?;
         let stdout = String::from_utf8_lossy(&output.stdout);
 
-        candidates.extend(stdout.lines().map(|line| {
-            let (id, desc) = split_help_text(line);
-            CompletionCandidate::new(id)
-                .help(desc)
-                .display_order(Some(CHANGE_ID))
-        }));
+        candidates.extend(
+            stdout
+                .lines()
+                .map(split_help_text)
+                .filter(|(id, _)| id.starts_with(match_prefix))
+                .map(|(id, desc)| {
+                    CompletionCandidate::new(id)
+                        .help(desc)
+                        .display_order(Some(CHANGE_ID))
+                }),
+        );
 
         // revset aliases
 
         let revset_aliases = load_revset_aliases(&Ui::null(), settings.config())?;
         let mut symbol_names: Vec<_> = revset_aliases.symbol_names().collect();
         symbol_names.sort();
-        candidates.extend(symbol_names.into_iter().map(|symbol| {
-            let (_, defn) = revset_aliases.get_symbol(symbol).unwrap();
-            CompletionCandidate::new(symbol)
-                .help(Some(defn.into()))
-                .display_order(Some(REVSET_ALIAS))
-        }));
+        candidates.extend(
+            symbol_names
+                .into_iter()
+                .filter(|symbol| symbol.starts_with(match_prefix))
+                .map(|symbol| {
+                    let (_, defn) = revset_aliases.get_symbol(symbol).unwrap();
+                    CompletionCandidate::new(symbol)
+                        .help(Some(defn.into()))
+                        .display_order(Some(REVSET_ALIAS))
+                }),
+        );
 
         Ok(candidates)
     })
 }
 
 pub fn mutable_revisions() -> Vec<CompletionCandidate> {
-    revisions(Some("mutable()"))
+    revisions(Some("mutable()"), "")
 }
 
 pub fn all_revisions() -> Vec<CompletionCandidate> {
-    revisions(None)
+    revisions(None, "")
+}
+
+fn revset_expression_multi(
+    current: &std::ffi::OsStr,
+    revisions_filter: Option<&str>,
+) -> Vec<CompletionCandidate> {
+    let Some(current) = current.to_str() else {
+        return Vec::new();
+    };
+    let revset_string = strip_shell_quote(current);
+    let match_prefix = find_trailing_partial_name(revset_string).unwrap_or(revset_string);
+    let candidates = revisions(revisions_filter, match_prefix);
+    if match_prefix.len() == current.len() {
+        candidates
+    } else {
+        let prepend = &current[..current.len() - match_prefix.len()];
+        candidates
+            .into_iter()
+            .map(|candidate| candidate.add_prefix(prepend))
+            .collect()
+    }
+}
+
+fn strip_shell_quote(current: &str) -> &str {
+    current
+        .strip_prefix('\'')
+        .or_else(|| current.strip_prefix('"'))
+        .unwrap_or(current)
+}
```

Comment:
Looking more into this, I think this is a mistake on `clap_complete`. I'll submit a patch there so that `current` is always an unquoted string, similar to what shell completion functions receive.